streaming/receiver.c

```diff
@@ -51,47 +51,235 @@ static void rrdpush_receiver_thread_cleanup(void *ptr) {
-               cd->version);
-        return PARSER_RC_OK;    // Ignore error and continue stream
+void send_replication_req(RRDHOST *host, char *st_id, time_t start, time_t end) {
+    char message[RRD_ID_LENGTH_MAX+60];
+    sprintf(message,"REPLICATE %s %ld %ld\n", st_id, start, end);
+    int ret;
+    debug(D_STREAM, "Replicate command: %s",message);
+#ifdef ENABLE_HTTPS
+    SSL *conn = host->stream_ssl.conn ;
+    if(conn && !host->stream_ssl.flags) {
+        ret = SSL_write(conn, message, strlen(message));
+    } else {
+        ret = send(host->receiver->fd, message, strlen(message), MSG_DONTWAIT);
     }
-    if (remote_time_txt && *remote_time_txt) {
-        remote_time = str2ull(remote_time_txt);
-        time_t now = now_realtime_sec(), prev = rrdhost_last_entry_t(host);
-        time_t gap = 0;
-        if (prev == 0)
-            info("STREAM %s from %s: Initial connection (no gap to check), remote=%ld local=%ld slew=%ld",
-                 host->hostname, cd->cmd, remote_time, now, now-remote_time);
-        else {
-            gap = now - prev;
-            info("STREAM %s from %s: Checking for gaps... remote=%ld local=%ld..%ld slew=%ld  %ld-sec gap",
-                 host->hostname, cd->cmd, remote_time, prev, now, remote_time - now, gap);
+#else
+    ret = send(host->receiver->fd, message, strlen(message), MSG_DONTWAIT);
+#endif
+    if (ret != (int)strlen(message))
+        error("Failed to send replication request for %s!", st_id);
+}
+
+static void skip_gap(RRDSET *st, time_t first_t, time_t last_t) {
+    // dbengine should handle time jump
+    if (st->rrd_memory_mode != RRD_MEMORY_MODE_DBENGINE) {
+        rrdset_rdlock(st);
+        RRDDIM *rd;
+        rrddim_foreach_read(rd, st) {
+            long current_entry = st->current_entry;
+            debug(D_REPLICATION, "Legacy-mode on %s.%s filling empty slots %ld-%ld.",
+                  st->name, rd->name, (long)first_t, (long)last_t);
+            for (time_t gap_t = first_t; gap_t <= last_t; gap_t += st->update_every) {
+                rd->values[current_entry] = SN_EMPTY_SLOT;
+                current_entry = ((current_entry + 1) >= st->entries) ? 0 : current_entry + 1;
+            }
         }
-        char message[128];
-        sprintf(message,"REPLICATE %ld %ld\n", remote_time - gap, remote_time);
-        int ret;
-#ifdef ENABLE_HTTPS
-        SSL *conn = host->stream_ssl.conn ;
-        if(conn && !host->stream_ssl.flags) {
-            ret = SSL_write(conn, message, strlen(message));
+        rrdset_unlock(st);
+    }
+    st->last_updated.tv_sec = last_t;
+}
+
+
+PARSER_RC streaming_rep_begin(char **words, void *user_v, PLUGINSD_ACTION *plugins_action) {
+    PARSER_USER_OBJECT *user = user_v;
+    UNUSED(plugins_action);
+    char *id = words[1];
+    char *start_txt = words[2];
+    char *first_txt = words[3];
+    char *end_txt = words[4];
+    if (!id || !start_txt || !end_txt)
+        goto disable;
+
+    RRDSET *st = NULL;
```

Comment:
you are setting `NULL` for a variable that is assigned in the next line with a new value, so this value is never used.

Reply:
Fixed.